### DISCUSSION TITLE
How to make a plugin that make Parcel build a website from a bunch of Markdown files

### DISCUSSION THREAD

**Discussion by guoyunhe** [parcel-bundler/parcel#8896] [Q&A]
You may know [docusaurus](https://docusaurus.io/), a document generator based on webpack. I am trying to make something similar but simpler.

In a project, we have some markdown files:

```
README.md
CHANGELOG.md
docs/api.md
docs/cli.md
docs/get-started.md
docs/faq.md
```

And the Parcel plugin should scan `*.md` and `docs/*.md`. If parcel only supports `*.html` as entries, then the plugin must generates `*.html` entry.

My questions are:

1. How to run Parcel without an entry?
2. How can a plugin generate entry HTML file?

**Comment by mischnic** [comment]
1. Not really, but what you want instead is `parcel build 'docs/**/*.md' README.md` (list all files, also using globs)
2. You need a transformer that takes Markdown and outputs HTML, which is very easy to do but there's no such official plugin.

A heads-up: these two points are easy to solve, but the hard parts will be stuff like a sidebar or an RSS feed (so a page which lists multiple/all files in the build).

**Reply by guoyunhe** [reply]
Thanks for the info. I prefer to transform Markdown files to React components (use Parcel's MDX plugin). And the plugin should glue up these React components into a React app. So it need to generates:

```
index.html
index.js
App.js <- contains all necessary layouts and routes
```

If Parcel has to take an entry before executing plugins, I think I have to write my own scripts to create these missing files and then call Parcel?

**Reply by mischnic** [reply]
If you want to build multiple static HTML pages from your Markdown (like Eleventy), then you need what I wrote earlier.
If you want to create a SPA, then maybe do `import * as pages from "./docs/**/*.mdx"` in JS (with the [glob resolver](https://parceljs.org/features/dependency-resolution/#glob-specifiers))?

**Reply by guoyunhe** [reply]
Yes, I think I need some glob matching. And this should be a configurable option. So users can choose what files they want to include. It is still an early idea. I can share more when I make some progress.

**Reply by TechQuery** [reply]
> 1. Not really, but what you want instead is `parcel build 'docs/**/*.md' README.md` (list all files, also using globs)
> 2. You need a transformer that takes Markdown and outputs HTML, which is very easy to do but there's no such official plugin.
> 
> A heads-up: these two points are easy to solve, but the hard parts will be stuff like a sidebar or an RSS feed (so a page which lists multiple/all files in the build).

@mischnic I try to build a SSG project with Parcel 2 glob entries, but nothing outputs with no error: EasyWebApp/mark-wiki#1

**Reply by valotas** [reply]
Had [something like that](https://github.com/valotas/valotas.com/tree/master/packages) working for my personal blog. The basic idea was to have a custom transformer that could read txt files. This file then can be a [`sitemap.txt`](https://github.com/valotas/valotas.com/blob/master/packages/web/src/sitemap.txt) containing all the different entry points you might need to create/publish and itself will be transformed to a "proper" [sitemap](https://valotas.com/sitemap.txt).

The entry point for something like the above would be a [transformer like this one](https://github.com/valotas/valotas.com/blob/master/packages/parcel-transformer-staticsite/src/StaticSiteTransformer.ts).

It needs some work to make it more abstract and have it working standalone though.
